feat: implement per-peer transaction relay tracking and metrics

deadmanoz opened the pull request:
- Add TransactionTracker with memory-bounded peer state tracking
- Track `INV`, `GETDATA`, and `TX` messages for unsolicited/unannounced detection
- Add per-peer metrics: `tx_unsolicited_by_peer`, `tx_unannounced_by_peer`, `tx_peer_last_activity`
- Implement peer cleanup after 30min inactivity (1000 txs/peer limit)
- Use dependency injection pattern with `Optional<&TransactionTracker>`
- Enable per-peer transaction behaviour analysis in Grafana
- Zero overhead when `tx_tracker` is None, easy testing and disabling

deadmanoz committed:
implement per-peer transaction relay tracking and metrics

- Add TransactionTracker with memory-bounded peer state tracking
- Track INV, GETDATA, and TX messages for unsolicited/unannounced detection
- Add per-peer metrics: tx_unsolicited_by_peer, tx_unannounced_by_peer, tx_peer_last_activity
- Implement peer cleanup after 30min inactivity (1000 txs/peer limit)
- Use dependency injection pattern with Optional<&TransactionTracker>
- Enable per-peer transaction behavior analysis in Grafana
- Zero overhead when tx_tracker is None, easy testing and disabling

deadmanoz commented:
@claude please review!

deadmanoz committed:
refactor: improve code quality in transaction tracking

- Add Default impl for PeerTxState to eliminate repeated String::new()
- Extract duplicated connection cleanup logic into helper function
- Replace .or_insert_with(|| PeerTxState::new(String::new())) with .or_default()

0xB10C commented on tools/metrics/src/metrics.rs:
```diff
@@ -332,6 +338,9 @@ impl Metrics {
+        icv!(tx_unsolicited_by_peer, "Number of unsolicited transactions received per peer", [LABEL_PEER_ID, LABEL_PEER_ADDR], registry);
+        icv!(tx_unannounced_by_peer, "Number of unannounced transaction requests per peer", [LABEL_PEER_ID, LABEL_PEER_ADDR], registry);
+        igv!(tx_peer_last_activity, "Timestamp of last transaction activity per peer", [LABEL_PEER_ID], registry);
 
```
I think having a label for the peer id and peer address isn't optimal:
1. These are both theoretically unbounded labels and Prometheus will store each combination as a new time series. See for example the warning about this in the Prometheus docs on labels https://prometheus.io/docs/practices/naming/#labels. For peer-observer, this will result in two problems:
    1. This will make the data served on /metrics a lot bigger, since Prometheus regularly fetches /metrics, it will increase traffic by a lot. There are some high-cardinallity metrics already which I've been thinking about removing. My current workaround is to restart the metrics tool every 48h, which isn't optimal
    2. The Prometheus database will take up significantly more disk space storing all peer-id and peer-addr combinations as separate time series. Displaying these in Grafana might also be slow, since Prometheus will need to fetch a lot of data first.


My suggestion would be to only keep track of the total number of unsolicited and unannounced in the metrics tool:

```suggestion
        ic!(p2p_tx_unsolicited_total, "Number of unsolicited transactions received.", registry);
        ic!(p2p_tx_unannounced_total, "Number of unannounced transaction requests", registry);
```

But I agree that it's interesting to see who is the source of the unsolicited and unannounced. I wonder the future alerts tool (https://github.com/0xB10C/peer-observer/issues/185) could have a metric for it? Not sure yet.

deadmanoz replied:
Ahhh I did do a "global"/total version initially, but then went the per-peer route to better align with `txrelay`.

I see what you're saying - once prometheus has scraped them, its now prometheus's problem, no matter how much cleanup we do on the metrics producer side..

deadmanoz replied:
@0xB10C How about a top-K style approach?

We lose information about which peer is responsible on the prometheus side, but we avoid the cardinality problem and still retain a bit more granular information about what the top-K peers are doing? And we can still get the totals by having one of the K being the "other" bucket?
 
I feel like the totals are too blunt. But then again I've only started dipping my toes into this world 😬

0xB10C replied:
How would top-K style look as a prometheus metric? I'm not sure I understand. What data would it contain? Similar to a [prometheus histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)?

0xB10C replied:
Maybe it's worthwhile to approach this from the other direction: What do we want to see in a Grafana dashboard? 

My initial answer would have been: a rate of unsolicted tx and a rate of unnanounced transactions a node receives per time interval. There will be some background activity, but if someone starts pushing out a lot of e.g. unsolicited transactions, we see a spike and can look into it (websocket tool, debug logs, possibly alerts tool, ...). I think this would address https://github.com/0xB10C/peer-observer/issues/182

0xB10C replied:
Thinking about it a bit more: Would k-top mean: We choose e.g. k=5 and have a metric just for 5 peers with most unsolicted/unnanounced and report for these IP, peer id, ...?

0xB10C replied:
> Maybe it's worthwhile to approach this from the other direction: What do we want to see in a Grafana dashboard?

Another thing that might be interesting: Number of peers with at least one unsolicited / unannounced. This would show if this is something that peers do regularly or only some do it. Would also show if all of a sudden many of our peers started to publish unsolicited.

deadmanoz committed:
fix: correct type reference in cleanup_peer_tracking function

Change `Option<&crate::TxTracker>` to `Option<&TransactionTracker>`
to match the actual imported type name.

deadmanoz committed:
feat: add Prometheus registry cleanup for disconnected peers

- Remove per-peer transaction metrics when peers disconnect
- Prevents unbounded cardinality growth from peer churn
- Reduces storage from ~12GB to ~161MB over 2-week retention
- Cleanup includes tx_unsolicited_by_peer, tx_unannounced_by_peer, tx_peer_last_activity

deadmanoz committed:
refactor: remove Option wrapper and rename TransactionTracker

- Remove unnecessary Option wrapper from TransactionTracker parameters
- All usage sites always provide the tracker, making Option unnecessary
- Rename TransactionTracker to TxRelayTracker for better clarity
- Update all function signatures and implementations
- Simplify code by removing redundant `if let Some(tracker)` checks

deadmanoz committed:
fix: get peer address before removing peer in cleanup logic

Critical bug fix: retrieve peer address before calling remove_peer(),
otherwise get_peer_address() returns None and metrics cleanup never executes.
This ensures per-peer metrics are properly removed from Prometheus registry.
